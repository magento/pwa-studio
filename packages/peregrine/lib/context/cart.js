import React, {
    createContext,
    useContext,
    useEffect,
    useMemo,
    useCallback
} from 'react';
import { connect } from 'react-redux';
import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';

import { useAwaitQuery } from '@magento/peregrine/lib/hooks/useAwaitQuery';
import actions from '../store/actions/cart/actions';
import * as asyncActions from '../store/actions/cart/asyncActions';
import bindActionCreators from '../util/bindActionCreators';
import { useEventListener } from '../hooks/useEventListener';
import BrowserPersistence from '../util/simplePersistence';

const CartContext = createContext();

const isCartEmpty = cart =>
    !cart || !cart.details.items || cart.details.items.length === 0;

const getTotalQuantity = items =>
    items.reduce((total, item) => total + item.quantity, 0);

const CartContextProvider = props => {
    const { actions, asyncActions, cartState, children } = props;

    // Make deeply nested details easier to retrieve and provide empty defaults
    const derivedDetails = useMemo(() => {
        if (isCartEmpty(cartState)) {
            return {
                currencyCode: 'USD',
                numItems: 0,
                subtotal: 0
            };
        } else {
            return {
                currencyCode: cartState.details.prices.grand_total.currency,
                numItems: getTotalQuantity(cartState.details.items),
                subtotal: cartState.details.prices.grand_total.value
            };
        }
    }, [cartState]);

    const cartApi = useMemo(
        () => ({
            actions,
            ...asyncActions
        }),
        [actions, asyncActions]
    );

    const contextValue = useMemo(() => {
        const derivedCartState = {
            ...cartState,
            isEmpty: isCartEmpty(cartState),
            derivedDetails
        };

        return [derivedCartState, cartApi];
    }, [cartApi, cartState, derivedDetails]);

    const [fetchCartId] = useMutation(CREATE_CART_MUTATION);
    const fetchCartDetails = useAwaitQuery(CART_DETAILS_QUERY);

    // Storage listener to force a state update if cartId changes from another browser tab.
    const storageListener = useCallback(() => {
        const storage = new BrowserPersistence();
        const currentCartId = storage.getItem('cartId');
        const { cartId } = cartState;
        if (cartId && currentCartId && cartId !== currentCartId) {
            globalThis.location && globalThis.location.reload();
        }
    }, [cartState]);

    useEventListener(globalThis, 'storage', storageListener);

    useEffect(() => {
        // cartApi.getCartDetails initializes the cart if there isn't one.
        cartApi.getCartDetails({
            fetchCartId,
            fetchCartDetails
        });
    }, [cartApi, fetchCartDetails, fetchCartId]);

    return (
        <CartContext.Provider value={contextValue}>
            {children}
        </CartContext.Provider>
    );
};

const mapStateToProps = ({ cart }) => ({ cartState: cart });

const mapDispatchToProps = dispatch => ({
    actions: bindActionCreators(actions, dispatch),
    asyncActions: bindActionCreators(asyncActions, dispatch)
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CartContextProvider);

export const useCartContext = () => useContext(CartContext);

/**
 * We normally do not keep GQL queries in Peregrine. All components should pass
 * queries to talons/hooks. This is an exception to the rule because it would
 * be unecessarily complex to pass these queries to the context provider.
 */
const CREATE_CART_MUTATION = gql`
    mutation createCart {
        cartId: createEmptyCart
    }
`;

const CART_DETAILS_QUERY = gql`
    query checkUserIsAuthed($cartId: String!) {
        cart(cart_id: $cartId) {
            # The purpose of this query is to check that the user is authorized
            # to query on the current cart. Just fetch "id" to keep it small.
            id
        }
    }
`;
