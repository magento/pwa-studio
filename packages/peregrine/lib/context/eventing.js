import React, {
    createContext,
    useCallback,
    useContext,
    useMemo,
    useRef
} from 'react';
import { Observable } from 'zen-observable-ts';

const EVENT_NAME = 'eventing';
const EventingContext = createContext();

const EventingContextProvider = props => {
    const { children } = props;
    const cacheRef = useRef([]);

    const state = useMemo(
        () =>
            new Observable(observer => {
                for (const element of cacheRef.current) {
                    observer.next(element);
                }

                document.addEventListener(EVENT_NAME, event => {
                    observer.next(event.detail);
                });
            }),
        []
    );

    const dispatch = useCallback(detail => {
        const event = new CustomEvent(EVENT_NAME, { detail });

        cacheRef.current.push(detail);
        document.dispatchEvent(event);
    }, []);

    const contextValue = useMemo(
        () => [
            state,
            {
                dispatch,
                subscribe: state.subscribe
            }
        ],
        [dispatch, state]
    );

    return (
        <EventingContext.Provider value={contextValue}>
            {children}
        </EventingContext.Provider>
    );
};

export default EventingContextProvider;
export const useEventingContext = () => useContext(EventingContext);
