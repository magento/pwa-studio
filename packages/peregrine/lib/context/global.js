import React, {
    createContext,
    useContext,
    useEffect,
    useMemo,
    useState
} from 'react';
import { useHistory } from 'react-router-dom';
import { useApolloClient } from '@apollo/client';

const StaticContext = createContext();

/**
 *
 * @param {object} props
 * @param {string} props.origin
 * @param {string} props.url
 * @param {object} props.dom
 * @returns
 */
const GlobalContextProvider = props => {
    const { url, origin, dom } = props;
    const navigated = useHistoryListener();
    const client = useApolloClient();

    const contextValue = useMemo(() => {
        const fullUrl = IS_BROWSER ? globalThis.location.pathname : url;
        return {
            url: fullUrl,
            origin: IS_BROWSER ? globalThis.location.origin : origin,
            href: IS_BROWSER ? globalThis.location.href : `${origin}${fullUrl}`,
            dom: IS_BROWSER ? globalThis : dom,
            // While there was no navigation, the current root content is server rendered
            isServerContent: SSR_ENABLED ? IS_SERVER || !navigated : false,
            // disableNetworkFetches has to be set to false after hydration is complete
            isHydrated: SSR_ENABLED
                ? IS_BROWSER && !client.disableNetworkFetches
                : true
        };
    }, [client.disableNetworkFetches, dom, navigated, origin, url]);

    return (
        <StaticContext.Provider value={contextValue}>
            {props.children}
        </StaticContext.Provider>
    );
};

/**
 *
 * @returns {{origin: string, dom: window, url: string, href: string, isServerContent: boolean, isHydrated: boolean }}
 */
export const useGlobalContext = () => useContext(StaticContext);

/**
 * Hook that tells you if there has been any navigation since the app was loaded
 *
 * @returns {boolean}
 */
const useHistoryListener = () => {
    const [navigated, setNavigated] = useState(false);

    const history = useHistory();

    const handleNavigated = () => {
        setNavigated(true);
    };

    useEffect(() => {
        let unlisten;

        if (!navigated) {
            unlisten = history.listen(() => {
                handleNavigated();
            });
        }

        return () => {
            if (unlisten) {
                unlisten();
            }
        };
    }, [history, navigated]);

    return navigated;
};

export default GlobalContextProvider;
