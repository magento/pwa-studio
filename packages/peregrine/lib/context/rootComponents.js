import React, { createContext, useContext, useState } from 'react';

const RootComponentsContext = createContext();

const RootComponentsProvider = props => {
    const { children } = props;
    const state = useState(new Map());

    return (
        <RootComponentsContext.Provider value={state}>
            {children}
        </RootComponentsContext.Provider>
    );
};

export default RootComponentsProvider;
export const useRootComponents = () => useContext(RootComponentsContext);
