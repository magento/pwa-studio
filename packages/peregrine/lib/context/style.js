import React, {
    createContext,
    useCallback,
    useContext,
    useEffect,
    useMemo,
    useState
} from 'react';
import mergeClasses from '@magento/peregrine/lib/util/shallowMerge';

const StyleContext = createContext();
const isServer = !globalThis.document;

/**
 * Client-side. Injects rulesets into a StyleSheet element, returning a cleanup
 * function that can remove the rulesets.
 *
 * @param {object} cssModule
 */
const insertModule = cssModule => {
    if (typeof cssModule._insertCss === 'function') {
        cssModule._insertCss();
    }
};

/**
 * Server-side. Creates a callback that adds rulesets to a global Set.
 *
 * @param {Set} initialState
 */
const addModule = initialState => cssModule => {
    if (typeof cssModule._getCss === 'function') {
        initialState.add(cssModule._getCss());
    }
};

const StyleContextProvider = props => {
    const { children, initialState } = props;

    const api = useMemo(
        () => (isServer ? addModule(initialState) : insertModule),
        [initialState]
    );

    return (
        <StyleContext.Provider value={api}>{children}</StyleContext.Provider>
    );
};

export default StyleContextProvider;

export const useStyle = (cssModule, ...overrides) => {
    const insertCss = useContext(StyleContext);
    const [classes, setClasses] = useState(() =>
        mergeClasses(cssModule, ...overrides)
    );

    // this effect always runs, since rest args are always a new array
    useEffect(() => {
        // even the override objects are not memoized, typically
        // so it's easiest to just merge again
        const nextClasses = mergeClasses(cssModule, ...overrides);

        // and then compare the results to see if anything changed
        if (hasChanged(classes, nextClasses)) {
            // and then update if something did change
            setClasses(nextClasses);
        }
    }, [classes, cssModule, overrides]);

    // only recreate the callback when the classes have changed
    const runInsert = useCallback(() => {
        try {
            // TODO: maybe throw an error instead of failing silently?
            // unit tests would need to mock this hook, though
            if (insertCss) {
                insertCss(classes);
            }
        } catch (error) {
            console.error('could not insert css:', classes);
        }
    }, [classes, insertCss]);

    // only run the effect when the callback has been recreated
    useEffect(() => {
        // React hooks must run the same number of times and in the same order
        // so any conditionals belong inside the effect
        if (!isServer) {
            runInsert();
        }
    }, [runInsert]);

    // React ignores effects on the server, so run this one during render
    // even though it's an antipattern on the client
    if (isServer) {
        runInsert();
    }

    return classes;
};

function hasChanged(prev, next) {
    const prevEntries = Object.entries(prev);
    const nextEntries = Object.entries(next);
    const count = Math.max(prevEntries.length, nextEntries.length);
    let hasChanged = false;

    if (prevEntries.length !== nextEntries.length) {
        return true;
    }

    for (let index = 0; index < count; index++) {
        const [prevKey, prevValue] = prevEntries[index];
        const [nextKey, nextValue] = nextEntries[index];

        if (prevKey !== nextKey || prevValue !== nextValue) {
            hasChanged = true;
            break;
        }
    }

    return hasChanged;
}
