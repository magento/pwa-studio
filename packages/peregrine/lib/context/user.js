import React, { createContext, useContext, useEffect, useMemo } from 'react';
import { connect } from 'react-redux';

import actions from '../store/actions/user/actions';
import * as asyncActions from '../store/actions/user/asyncActions';
import bindActionCreators from '../util/bindActionCreators';
import BrowserPersistence from '../util/simplePersistence';

const UserContext = createContext();

const UserContextProvider = props => {
    const { actions, asyncActions, children, userState } = props;

    const userApi = useMemo(
        () => ({
            actions,
            ...asyncActions
        }),
        [actions, asyncActions]
    );

    const contextValue = useMemo(() => [userState, userApi], [
        userApi,
        userState
    ]);

    useEffect(() => {
        // check if the user's token is not expired
        const storage = new BrowserPersistence();
        const item = storage.getRawItem('signin_token');

        if (item) {
            const { ttl, timeStored } = JSON.parse(item);
            const now = Date.now();

            // if the token's TTYL has expired, we need to sign out
            if (ttl && now - timeStored > ttl * 1000) {
                asyncActions.signOut();
            }
        }
    }, [asyncActions]);

    return (
        <UserContext.Provider value={contextValue}>
            {children}
        </UserContext.Provider>
    );
};

const mapStateToProps = ({ user }) => ({ userState: user });

const mapDispatchToProps = dispatch => ({
    actions: bindActionCreators(actions, dispatch),
    asyncActions: bindActionCreators(asyncActions, dispatch)
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(UserContextProvider);

/**
 * @typedef {Object} UserState
 *
 * @property {CurrentUser} currentUser Current user details
 * @property {Error} getDetailsError Get Details call related error
 * @property {Boolean} isGettingDetails Boolean if true indicates that user details are being fetched. False otherwise.
 * @property {Boolean} isResettingPassword Deprecated
 * @property {Boolean} isSignedIn Boolean if true indicates that the user is signed in. False otherwise.
 * @property {Error} resetPasswordError Deprecated
 *
 */

/**
 * @typedef {Object} CurrentUser
 *
 * @property {String} email Current user's email
 * @property {String} firstname Current user's first name
 * @property {String} lastname Current user's last name
 */

/**
 * @typedef {Object} UserActions
 *
 * @property {Function} clearToken Callback to clear user token in browser persistence storage
 * @property {Function} getUserDetails Callback to get user details
 * @property {Function} resetPassword Deprecated
 * @property {Function} setToken Callback to set user token in browser persistence storage
 * @property {Function} signOut Callback to sign the user out
 */

/**
 * @returns {[UserState, UserActions]}
 */
export const useUserContext = () => useContext(UserContext);
