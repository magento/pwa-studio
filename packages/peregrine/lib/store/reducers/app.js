import { handleActions } from 'redux-actions';

import actions from '../actions/app';

export const name = 'app';

// as far as the server is concerned, the app is always online
const isServer = !globalThis.navigator;
const isOnline = !isServer && navigator.onLine;
const hasBeenOffline = !isServer && !navigator.onLine;

const initialState = {
    drawer: null,
    hasBeenOffline,
    isOnline,
    isPageLoading: false,
    overlay: false,
    pending: {},
    searchOpen: false
};

const reducerMap = {
    [actions.toggleDrawer]: (state, { payload }) => {
        return {
            ...state,
            drawer: payload,
            overlay: !!payload
        };
    },
    [actions.toggleSearch]: state => {
        return {
            ...state,
            searchOpen: !state.searchOpen
        };
    },
    [actions.setOnline]: state => {
        return {
            ...state,
            isOnline: true
        };
    },
    [actions.setOffline]: state => {
        return {
            ...state,
            isOnline: false,
            hasBeenOffline: true
        };
    },
    [actions.setPageLoading]: (state, { payload }) => {
        return {
            ...state,
            isPageLoading: !!payload
        };
    }
};

export default handleActions(reducerMap, initialState);
