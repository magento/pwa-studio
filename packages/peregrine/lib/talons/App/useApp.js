import { useCallback, useEffect, useMemo } from 'react';
import { useHistory } from 'react-router-dom';

import errorRecord from '@magento/peregrine/lib/util/createErrorRecord';
import { useAppContext } from '@magento/peregrine/lib/context/app';

const dismissers = new WeakMap();

// Memoize dismisser funcs to reduce re-renders from func identity change.
const getErrorDismisser = (error, onDismissError) => {
    return dismissers.has(error)
        ? dismissers.get(error)
        : dismissers.set(error, () => onDismissError(error)).get(error);
};

/**
 * Talon that handles effects for App and returns props necessary for rendering
 * the app.
 *
 * @param {Function} props.handleError callback to invoke for each error
 * @param {Function} props.handleIsOffline callback to invoke when the app goes offline
 * @param {Function} props.handleIsOnline callback to invoke wen the app goes online
 * @param {Function} props.handleHTMLUpdate callback to invoke when a HTML update is available
 * @param {Function} props.markErrorHandled callback to invoke when handling an error
 * @param {Function} props.renderError an error that occurs during rendering of the app
 * @param {Function} props.unhandledErrors errors coming from the error reducer
 *
 * @returns {{
 *  hasOverlay: boolean
 *  handleCloseDrawer: function
 * }}
 */
export const useApp = props => {
    const {
        handleError,
        handleIsOffline,
        handleIsOnline,
        markErrorHandled,
        renderError,
        unhandledErrors
    } = props;
    const history = useHistory();

    const reload = useCallback(() => {
        if (process.env.NODE_ENV !== 'development') {
            history.go(0);
        }
    }, [history]);

    const renderErrors = useMemo(
        () =>
            renderError
                ? [
                      errorRecord(
                          renderError,
                          globalThis,
                          useApp,
                          renderError.stack
                      )
                  ]
                : [],
        [renderError]
    );

    const errors = renderError ? renderErrors : unhandledErrors;
    const handleDismissError = renderError ? reload : markErrorHandled;

    // Only add toasts for errors if the errors list changes. Since `addToast`
    // and `toasts` changes each render we cannot add it as an effect dependency
    // otherwise we infinitely loop.
    useEffect(() => {
        for (const { error, id, loc } of errors) {
            handleError(
                error,
                id,
                loc,
                getErrorDismisser(error, handleDismissError)
            );
        }
    }, [errors, handleDismissError, handleError]);

    const [appState, appApi] = useAppContext();
    const { closeDrawer } = appApi;
    const { hasBeenOffline, isOnline, overlay } = appState;

    useEffect(() => {
        if (hasBeenOffline) {
            if (isOnline) {
                handleIsOnline();
            } else {
                handleIsOffline();
            }
        }
    }, [handleIsOnline, handleIsOffline, hasBeenOffline, isOnline]);

    const handleCloseDrawer = useCallback(() => {
        closeDrawer();
    }, [closeDrawer]);

    return {
        hasOverlay: !!overlay,
        handleCloseDrawer
    };
};
