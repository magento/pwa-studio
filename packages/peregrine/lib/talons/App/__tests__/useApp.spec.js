import React from 'react';
import { useHistory } from 'react-router-dom';
import { createTestInstance } from '@magento/peregrine';
import errorRecord from '@magento/peregrine/lib/util/createErrorRecord';
import { act } from 'react-test-renderer';

import { useApp } from '../useApp';

import { useAppContext } from '@magento/peregrine/lib/context/app';

jest.mock('react-router-dom', () => ({
    useHistory: jest.fn()
}));
const reload = jest.fn();
useHistory.mockImplementation(() => ({
    go: reload
}));

jest.mock('@magento/peregrine/lib/context/app', () => {
    const state = {
        drawerClosed: false,
        isOnline: true,
        hasBeenOffline: false,
        overlay: false
    };

    const api = {
        closeDrawer: jest.fn(() => {
            state.drawerClosed = true;
        }),
        setOnline: jest.fn(isOnline => {
            state.isOnline = isOnline;
        }),
        setHasBeenOffline: jest.fn(hasBeenOffline => {
            state.hasBeenOffline = hasBeenOffline;
        })
    };

    const useAppContext = jest.fn(() => [state, api]);

    return { useAppContext };
});

const log = jest.fn();
const Component = props => {
    const talonProps = useApp(props);
    log(talonProps);

    return <i />;
};

const handleError = jest.fn((error, id, loc, dismisser) => {
    dismisser();
});
const handleIsOffline = jest.fn();
const handleIsOnline = jest.fn();
const handleHTMLUpdate = jest.fn();
const markErrorHandled = jest.fn();
const renderError = undefined;
const unhandledErrors = [];
const mockProps = {
    handleError,
    handleIsOffline,
    handleIsOnline,
    handleHTMLUpdate,
    markErrorHandled,
    renderError,
    unhandledErrors
};

test('returns the correct shape', () => {
    createTestInstance(<Component {...mockProps} />);

    const talonProps = log.mock.calls[0][0];

    expect(talonProps).toMatchSnapshot();
});

test('handle render error', () => {
    const error = new Error('Render error');

    const props = {
        ...mockProps,
        renderError: error
    };

    createTestInstance(<Component {...props} />);

    expect(handleError).toHaveBeenCalledTimes(1);
    expect(reload).toHaveBeenCalledTimes(1);
});

test('handles unhandled errors', () => {
    const error1 = new Error('Error 1');
    const error2 = new Error('Error 2');
    const error3 = new Error('Error 3');

    const props = {
        ...mockProps,
        unhandledErrors: [
            errorRecord(error1, window, useApp, error1.stack),
            errorRecord(error2, window, useApp, error2.stack),
            errorRecord(error3, window, useApp, error3.stack)
        ]
    };

    const component = createTestInstance(<Component {...props} />);

    expect(markErrorHandled).toHaveBeenCalledTimes(3);
    expect(handleError).toHaveBeenCalledTimes(3);

    // Re-handle same error
    act(() => {
        const newProps = {
            ...mockProps,
            unhandledErrors: [errorRecord(error1, window, useApp, error1.stack)]
        };
        markErrorHandled.mockClear();
        handleError.mockClear();
        component.update(<Component {...newProps} />);
    });

    expect(markErrorHandled).toHaveBeenCalledTimes(1);
    expect(handleError).toHaveBeenCalledTimes(1);
});

test('handle render error in development environment', () => {
    const error = new Error('Render error');

    const props = {
        ...mockProps,
        renderError: error
    };

    process.env.NODE_ENV = 'development';

    createTestInstance(<Component {...props} />);

    expect(handleError).toHaveBeenCalledTimes(1);
    expect(reload).toHaveBeenCalledTimes(0);
});

test('handle drawer closing', () => {
    createTestInstance(<Component {...mockProps} />);

    const talonProps = log.mock.calls[0][0];

    const { handleCloseDrawer } = talonProps;

    const [appState] = useAppContext();

    expect(appState.drawerClosed).toBeFalsy();

    act(() => {
        handleCloseDrawer();
    });

    expect(appState.drawerClosed).toBeTruthy();
});

test('handle being offline to online event', () => {
    const [, appApi] = useAppContext();

    const { setOnline, setHasBeenOffline } = appApi;

    act(() => {
        setHasBeenOffline(true);
        setOnline(true);
        createTestInstance(<Component {...mockProps} />);
    });

    expect(handleIsOnline).toHaveBeenCalledTimes(1);
    expect(handleIsOffline).toHaveBeenCalledTimes(0);
});

test('handle being offline', () => {
    const [, appApi] = useAppContext();

    const { setOnline, setHasBeenOffline } = appApi;

    act(() => {
        setHasBeenOffline(true);
        setOnline(false);
        createTestInstance(<Component {...mockProps} />);
    });

    expect(handleIsOnline).toHaveBeenCalledTimes(0);
    expect(handleIsOffline).toHaveBeenCalledTimes(1);
});
