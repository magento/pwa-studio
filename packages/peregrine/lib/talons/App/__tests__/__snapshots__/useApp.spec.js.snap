// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns the correct shape 1`] = `
Object {
  "handleCloseDrawer": [Function],
  "hasOverlay": false,
}
`;
